// @awsui/components-react/test-utils/dom/button-dropdown/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var dom_1 = require("@awsui/test-utils-core/dom");
var styles_selectors_js_1 = require("../../../button-dropdown/styles.selectors.js");
var styles_selectors_js_2 = require("../../../internal/components/dropdown/styles.selectors.js");
var styles_selectors_js_3 = require("../../../button-dropdown/item-element/styles.selectors.js");
var styles_selectors_js_4 = require("../../../button-dropdown/category-elements/styles.selectors.js");
var styles_selectors_js_5 = require("../../../button/styles.selectors.js");
var ButtonDropdownWrapper = (function (_super) {
    __extends(ButtonDropdownWrapper, _super);
    function ButtonDropdownWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ButtonDropdownWrapper.prototype.findNativeButton = function () {
        return this.find("button." + styles_selectors_js_5.default.button);
    };
    ButtonDropdownWrapper.prototype.findOpenDropdown = function () {
        return (0, dom_1.createWrapper)().find("." + styles_selectors_js_2.default.dropdown + "[data-open=true]");
    };
    ButtonDropdownWrapper.prototype.findItemById = function (id) {
        var _a;
        var itemSelector = "." + styles_selectors_js_3.default['item-element'] + "[data-testid=\"" + id + "\"]";
        return ((_a = this.findOpenDropdown()) === null || _a === void 0 ? void 0 : _a.find(itemSelector)) || this.find(itemSelector);
    };
    ButtonDropdownWrapper.prototype.findExpandableCategoryById = function (id) {
        var _a;
        var expandableCategorySelector = "." + styles_selectors_js_4.default.expandable + "[data-testid=\"" + id + "\"]";
        return ((_a = this.findOpenDropdown()) === null || _a === void 0 ? void 0 : _a.find(expandableCategorySelector)) || this.find(expandableCategorySelector);
    };
    ButtonDropdownWrapper.prototype.findHighlightedItem = function () {
        var _a;
        var highlightedItemSelector = "." + styles_selectors_js_3.default['item-element'] + "." + styles_selectors_js_3.default.highlighted;
        return ((_a = this.findOpenDropdown()) === null || _a === void 0 ? void 0 : _a.find(highlightedItemSelector)) || this.find(highlightedItemSelector);
    };
    ButtonDropdownWrapper.prototype.findItems = function () {
        var _a;
        return ((_a = this.findOpenDropdown()) === null || _a === void 0 ? void 0 : _a.findAll("." + styles_selectors_js_3.default['item-element'])) || [];
    };
    ButtonDropdownWrapper.prototype.findDisabledReason = function () {
        return (0, dom_1.createWrapper)().find("[data-testid=\"button-dropdown-disabled-reason\"]");
    };
    ButtonDropdownWrapper.prototype.openDropdown = function () {
        this.findNativeButton().click();
    };
    ButtonDropdownWrapper.rootSelector = styles_selectors_js_1.default['button-dropdown'];
    __decorate([
        dom_1.usesDom
    ], ButtonDropdownWrapper.prototype, "openDropdown", null);
    return ButtonDropdownWrapper;
}(dom_1.ComponentWrapper));
exports.default = ButtonDropdownWrapper;
