// @awsui/components-react/internal/generated/custom-css-properties/index.js
var customCSSPropertiesMap = {
    offsetTop: '--awsui-offset-top-bq7ab8',
    offsetTopWithNotifications: '--awsui-offset-top-with-notifications-bq7ab8',
    contentGapLeft: '--awsui-content-gap-left-bq7ab8',
    contentGapRight: '--awsui-content-gap-right-bq7ab8',
    contentHeight: '--awsui-content-height-bq7ab8',
    defaultMaxContentWidth: '--awsui-default-max-content-width-bq7ab8',
    defaultMinContentWidth: '--awsui-default-min-content-width-bq7ab8',
    footerHeight: '--awsui-footer-height-bq7ab8',
    headerHeight: '--awsui-header-height-bq7ab8',
    layoutWidth: '--awsui-layout-width-bq7ab8',
    mainOffsetLeft: '--awsui-main-offset-left-bq7ab8',
    maxContentWidth: '--awsui-max-content-width-bq7ab8',
    minContentWidth: '--awsui-min-content-width-bq7ab8',
    notificationsHeight: '--awsui-notifications-height-bq7ab8',
    overlapHeight: '--awsui-overlap-height-bq7ab8',
    navigationWidth: '--awsui-navigation-width-bq7ab8',
    splitPanelReportedHeaderSize: '--awsui-split-panel-reported-header-size-bq7ab8',
    splitPanelReportedSize: '--awsui-split-panel-reported-size-bq7ab8',
    splitPanelMinWidth: '--awsui-split-panel-min-width-bq7ab8',
    splitPanelMaxWidth: '--awsui-split-panel-max-width-bq7ab8',
    toolsMaxWidth: '--awsui-tools-max-width-bq7ab8',
    toolsWidth: '--awsui-tools-width-bq7ab8',
    toolsAnimationStartingOpacity: '--awsui-tools-animation-starting-opacity-bq7ab8'
};
export default customCSSPropertiesMap;
