// @awsui/components-react/test-utils/dom/tabs/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var dom_1 = require("@awsui/test-utils-core/dom");
var styles_selectors_js_1 = require("../../../tabs/styles.selectors.js");
var TabsWrapper = (function (_super) {
    __extends(TabsWrapper, _super);
    function TabsWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TabsWrapper.prototype.findTabLinks = function () {
        return this.findAllByClassName(styles_selectors_js_1.default['tabs-tab-link']);
    };
    TabsWrapper.prototype.findTabLinkByIndex = function (index) {
        return this.find("." + styles_selectors_js_1.default['tabs-tab'] + ":nth-child(" + index + ") ." + styles_selectors_js_1.default['tabs-tab-link']);
    };
    TabsWrapper.prototype.findTabLinkById = function (id) {
        return this.find("." + styles_selectors_js_1.default['tabs-tab-link'] + "[data-testid=\"" + id + "\"]");
    };
    TabsWrapper.prototype.findActiveTab = function () {
        return this.find("." + styles_selectors_js_1.default['tabs-tab-active']);
    };
    TabsWrapper.prototype.findTabContent = function () {
        return this.find("." + styles_selectors_js_1.default['tabs-content-active']);
    };
    TabsWrapper.rootSelector = styles_selectors_js_1.default.root;
    return TabsWrapper;
}(dom_1.ComponentWrapper));
exports.default = TabsWrapper;
