// @awsui/components-react/wizard/wizard-navigation.js
import React from 'react';
import clsx from 'clsx';
import InternalLink from '../link/internal';
import InternalBox from '../box/internal';
import styles from './styles.css.js';
var Statuses;
(function (Statuses) {
    Statuses["Active"] = "active";
    Statuses["Unvisited"] = "unvisited";
    Statuses["Visited"] = "visited";
    Statuses["Next"] = "next";
})(Statuses || (Statuses = {}));
export default function Navigation(_a) {
    var activeStepIndex = _a.activeStepIndex, farthestStepIndex = _a.farthestStepIndex, allowSkipTo = _a.allowSkipTo, hidden = _a.hidden, i18nStrings = _a.i18nStrings, isVisualRefresh = _a.isVisualRefresh, onStepClick = _a.onStepClick, onSkipToClick = _a.onSkipToClick, steps = _a.steps;
    return (React.createElement("nav", { className: clsx(styles.navigation, hidden && styles.hidden, isVisualRefresh && styles.refresh) },
        React.createElement("ul", null, steps.map(function (step, index) {
            return isVisualRefresh ? (React.createElement(NavigationStepVisualRefresh, { i18nStrings: i18nStrings, index: index, key: index, onStepClick: onStepClick, onSkipToClick: onSkipToClick, status: getStatus(index), step: step })) : (React.createElement(NavigationStepClassic, { i18nStrings: i18nStrings, index: index, key: index, onStepClick: onStepClick, onSkipToClick: onSkipToClick, status: getStatus(index), step: step }));
        }))));
    function getStatus(index) {
        if (activeStepIndex === index) {
            return Statuses.Active;
        }
        if (farthestStepIndex >= index) {
            return Statuses.Visited;
        }
        if (allowSkipTo && canSkip(activeStepIndex, index)) {
            return Statuses.Next;
        }
        return Statuses.Unvisited;
    }
    function canSkip(fromIndex, toIndex) {
        for (var index = fromIndex + 1; index < toIndex; index++) {
            if (!steps[index].isOptional) {
                return false;
            }
        }
        return true;
    }
}
function NavigationStepVisualRefresh(_a) {
    var i18nStrings = _a.i18nStrings, index = _a.index, onStepClick = _a.onStepClick, onSkipToClick = _a.onSkipToClick, status = _a.status, step = _a.step;
    function handleStepInteraction() {
        if (status === Statuses.Visited) {
            onStepClick(index);
        }
        if (status === Statuses.Next) {
            onSkipToClick(index);
        }
    }
    var state = {
        active: 'active',
        unvisited: 'disabled',
        visited: 'enabled',
        next: 'enabled'
    }[status];
    return (React.createElement("li", { className: clsx(styles["" + state]) },
        React.createElement("hr", null),
        React.createElement("span", { className: clsx(styles.number) },
            i18nStrings.stepNumberLabel && i18nStrings.stepNumberLabel(index + 1),
            step.isOptional && React.createElement("i", null, " - " + i18nStrings.optional)),
        React.createElement("a", { "aria-current": status === Statuses.Active ? 'step' : undefined, "aria-disabled": status === Statuses.Unvisited ? 'true' : undefined, onClick: function (event) {
                event.preventDefault();
                handleStepInteraction();
            }, onKeyDown: function (event) {
                if (event.key === ' ' || event.key === 'Enter') {
                    event.preventDefault();
                }
                if (event.key === 'Enter') {
                    handleStepInteraction();
                }
            }, onKeyUp: function (event) {
                if (event.key === ' ') {
                    handleStepInteraction();
                }
            }, role: "button", tabIndex: status === Statuses.Visited || status === Statuses.Next ? 0 : undefined },
            React.createElement("div", { className: clsx(styles.circle) }),
            React.createElement("span", { className: clsx(styles.title) }, step.title))));
}
function NavigationStepClassic(_a) {
    var i18nStrings = _a.i18nStrings, index = _a.index, onStepClick = _a.onStepClick, onSkipToClick = _a.onSkipToClick, status = _a.status, step = _a.step;
    var spanClassName = clsx(styles['navigation-link'], status === Statuses.Active ? styles['navigation-link-active'] : styles['navigation-link-disabled']);
    return (React.createElement("li", { className: styles['navigation-link-item'] },
        React.createElement(InternalBox, { variant: "small", className: styles['navigation-link-label'], display: "block", margin: { bottom: 'xxs' } },
            i18nStrings.stepNumberLabel && i18nStrings.stepNumberLabel(index + 1),
            step.isOptional && React.createElement("i", null, " - " + i18nStrings.optional)),
        React.createElement("div", null, status === Statuses.Visited || status === Statuses.Next ? (React.createElement(InternalLink, { className: clsx(styles['navigation-link']), onFollow: function (evt) {
                evt.preventDefault();
                status === Statuses.Visited ? onStepClick(index) : onSkipToClick(index);
            } }, step.title)) : (React.createElement("span", { className: spanClassName, "aria-current": status === Statuses.Active ? 'step' : undefined, "aria-disabled": status === Statuses.Active ? undefined : 'true' }, step.title)))));
}
