// @awsui/components-react/top-navigation/use-top-navigation.js
import { __spreadArray } from "tslib";
import { useEffect, useMemo, useRef, useState } from 'react';
import mergeRefs from 'react-merge-refs';
import { useContainerBreakpoints, useContainerQuery } from '../internal/hooks/container-queries';
import styles from './styles.css.js';
var RESPONSIVENESS_BUFFER = 20;
export function useTopNavigation(_a) {
    var mainRef = _a.__internalRootRef, identity = _a.identity, search = _a.search, utilities = _a.utilities;
    var virtualRef = useRef(null);
    var _b = useContainerBreakpoints(['xxs', 's']), breakpoint = _b[0], breakpointRef = _b[1];
    var hasSearch = !!search;
    var hasTitleWithLogo = identity && !!identity.logo && !!identity.title;
    var responsiveStates = useMemo(function () {
        return generateResponsiveStateKeys(utilities, hasSearch, hasTitleWithLogo);
    }, [utilities, hasSearch, hasTitleWithLogo]);
    var _c = useContainerQuery(function () {
        var _a, _b, _c, _d;
        if (!(mainRef === null || mainRef === void 0 ? void 0 : mainRef.current) || !virtualRef.current) {
            return responsiveStates[0];
        }
        var sizeConfiguration = {
            hasSearch: !!search,
            availableWidth: getContentBoxWidth(mainRef.current.querySelector("." + styles['padding-box'])),
            utilitiesLeftPadding: parseFloat(getComputedStyle(virtualRef.current.querySelector("." + styles.utilities)).paddingLeft),
            fullIdentityWidth: virtualRef.current.querySelector("." + styles.identity).getBoundingClientRect().width,
            titleWidth: (_b = (_a = virtualRef.current.querySelector("." + styles.title)) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect().width) !== null && _b !== void 0 ? _b : 0,
            searchSlotWidth: (_d = (_c = virtualRef.current.querySelector("." + styles.search)) === null || _c === void 0 ? void 0 : _c.getBoundingClientRect().width) !== null && _d !== void 0 ? _d : 0,
            searchUtilityWidth: virtualRef.current.querySelector('[data-utility-special="search"]').getBoundingClientRect()
                .width,
            utilityWithLabelWidths: Array.prototype.slice
                .call(virtualRef.current.querySelectorAll("[data-utility-hide=\"false\"]"))
                .map(function (element) { return element.getBoundingClientRect().width; }),
            utilityWithoutLabelWidths: Array.prototype.slice
                .call(virtualRef.current.querySelectorAll("[data-utility-hide=\"true\"]"))
                .map(function (element) { return element.getBoundingClientRect().width; }),
            menuTriggerUtilityWidth: virtualRef.current
                .querySelector('[data-utility-special="menu-trigger"]')
                .getBoundingClientRect().width
        };
        return determineBestResponsiveState(responsiveStates, sizeConfiguration);
    }, [mainRef, search, responsiveStates]), responsiveState = _c[0], containerQueryRef = _c[1];
    var _d = useState(true), isSearchMinimized = _d[0], setSearchMinimized = _d[1];
    var isSearchExpanded = !isSearchMinimized && breakpoint !== 's' && hasSearch && (responsiveState === null || responsiveState === void 0 ? void 0 : responsiveState.hideSearch);
    useEffect(function () {
        if (!(responsiveState === null || responsiveState === void 0 ? void 0 : responsiveState.hideSearch)) {
            setSearchMinimized(true);
        }
    }, [responsiveState]);
    useEffect(function () {
        var _a, _b;
        if (isSearchExpanded) {
            (_b = (_a = mainRef === null || mainRef === void 0 ? void 0 : mainRef.current) === null || _a === void 0 ? void 0 : _a.querySelector("." + styles.search + " input")) === null || _b === void 0 ? void 0 : _b.focus();
        }
    }, [isSearchExpanded, mainRef]);
    return {
        ref: mergeRefs(__spreadArray(__spreadArray([], (mainRef ? [mainRef] : []), true), [containerQueryRef, breakpointRef], false)),
        virtualRef: virtualRef,
        responsiveState: responsiveState !== null && responsiveState !== void 0 ? responsiveState : responsiveStates[0],
        breakpoint: breakpoint !== null && breakpoint !== void 0 ? breakpoint : 'default',
        isSearchExpanded: !!isSearchExpanded,
        onSearchUtilityClick: function () { return setSearchMinimized(function (isSearchMinimized) { return !isSearchMinimized; }); }
    };
}
function getContentBoxWidth(element) {
    var style = getComputedStyle(element);
    return parseFloat(style.width) - parseFloat(style.paddingLeft) - parseFloat(style.paddingRight);
}
export function generateResponsiveStateKeys(utilities, canHideSearch, canHideTitle) {
    var states = [{}];
    if (utilities.some(function (utility) { return utility.text; })) {
        states.push({ hideUtilityText: true });
    }
    if (canHideSearch) {
        states.push({
            hideUtilityText: true,
            hideSearch: true
        });
    }
    var hiddenUtilties = [];
    for (var i = 0; i < utilities.length; i++) {
        if (!utilities[i].disableUtilityCollapse) {
            hiddenUtilties.push(i);
            states.push({
                hideUtilityText: true,
                hideSearch: canHideSearch || undefined,
                hideUtilities: hiddenUtilties.length > 0 ? hiddenUtilties.slice() : undefined
            });
        }
    }
    if (canHideTitle) {
        states.push({
            hideUtilityText: true,
            hideSearch: canHideSearch || undefined,
            hideUtilities: hiddenUtilties.length > 0 ? hiddenUtilties.slice() : undefined,
            hideTitle: true
        });
    }
    return states;
}
export function determineBestResponsiveState(possibleStates, sizes) {
    var hasSearch = sizes.hasSearch, availableWidth = sizes.availableWidth, utilitiesLeftPadding = sizes.utilitiesLeftPadding, fullIdentityWidth = sizes.fullIdentityWidth, titleWidth = sizes.titleWidth, searchSlotWidth = sizes.searchSlotWidth, searchUtilityWidth = sizes.searchUtilityWidth, utilityWithLabelWidths = sizes.utilityWithLabelWidths, utilityWithoutLabelWidths = sizes.utilityWithoutLabelWidths, menuTriggerUtilityWidth = sizes.menuTriggerUtilityWidth;
    var _loop_1 = function (state) {
        var searchWidth = hasSearch ? (state.hideSearch ? searchUtilityWidth : searchSlotWidth) : 0;
        var utilitiesWidth = (state.hideUtilityText ? utilityWithoutLabelWidths : utilityWithLabelWidths)
            .filter(function (_width, i) { return !state.hideUtilities || state.hideUtilities.indexOf(i) === -1; })
            .reduce(function (sum, width) { return sum + width; }, 0);
        var menuTriggerWidth = state.hideUtilities ? menuTriggerUtilityWidth : 0;
        var identityWidth = state.hideTitle ? fullIdentityWidth - titleWidth : fullIdentityWidth;
        var expectedInnerWidth = identityWidth + searchWidth + utilitiesLeftPadding + utilitiesWidth + menuTriggerWidth;
        if (expectedInnerWidth <= availableWidth - RESPONSIVENESS_BUFFER) {
            return { value: state };
        }
    };
    for (var _i = 0, possibleStates_1 = possibleStates; _i < possibleStates_1.length; _i++) {
        var state = possibleStates_1[_i];
        var state_1 = _loop_1(state);
        if (typeof state_1 === "object")
            return state_1.value;
    }
    return possibleStates[possibleStates.length - 1];
}
